test-data/unit/check-functions.test

```diff
@@ -2542,3 +2542,108 @@ def f() -> int: ...
 [file p/d.py]
 import p
 def f() -> int: ...
+
+[case testFactory]
+from typing import Type
+class A:
+    x = "foo"
+def f() -> Type[A]:
+    return A
+def g() -> A:
+    return A()
+
+f().x = "bar"
+f().x = 0 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+f()().x = "baz"
+f()().x = 1 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+g().x = ""
+g().x = 2 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+[case testFactoryWithInheritance]
+from typing import Type
+class A:
+    x = "foo"
+class B(A):
+    pass
+def f() -> Type[B]:
+    return B
+def g() -> B:
+    return B()
+
+f().x = "bar"
+f().x = 0 # E: Incompatible types in assignment (expression has type "int", base class "A" defined the type as "str")
+f()().x = "baz"
+f()().x = 1 # E: Incompatible types in assignment (expression has type "int", base class "A" defined the type as "str")
+g().x = ""
+g().x = 2 # E: Incompatible types in assignment (expression has type "int", base class "A" defined the type as "str")
+
+[case testRecursiveCallable]
+from typing import Optional
+class A:
+    x = 0 # Type: int
```

Comment:
* You don't need to use type comments in tests.
* This is not a type comment because of a typo (uppercase `T`)
* It is not needed here